### PR TITLE
src folder for code review

### PR DESCRIPTION
**Pull request by utsav-deep-rjil**

### CODE REVIEW

File: src/main/java/com/jcs/sbs/auth/Authorization.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            throw new RuntimeException("Unsupported protocol present in given url :" + url);
+        }
+        this.protocol = protocolAndHost.getProtocol();
+        this.host = protocolAndHost.getHost();
+        this.port = "";
+        if (host.indexOf(':') > 0) {
+            String[] parts = host.split(":");
+            this.host = parts[0];
+            this.port = parts[1];
+        }
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/auth/Authorization.java:71]: Better to throw an exception if host and port ar not valid. Do you want to consider this ?

File: src/main/java/com/jcs/sbs/auth/Authorization.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+     * Method to convert query params as java map to sorted query params as in
+     * format of request URLs
+     * 
+     * @param params
+     *            Query parameters as map
+     * @return Query parameters converted to UTF-8 string in format used in
+     *         request URLs
+     * @throws UnsupportedEncodingException
+     */
+    public String sortParams(Map<String, String> params) throws UnsupportedEncodingException {
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/auth/Authorization.java]: its probably Serializing the params so sort is not the right name here

File: src/main/java/com/jcs/sbs/auth/BasicJCSCredentials.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    /**
+     * constructor to initialize accessKey and secretKey
+     * 
+     * @param accessKey
+     *            Unique JCS Access Key of the account
+     * @param secretKey
+     *            JCS Secret Key of the account
+     */
+    public BasicJCSCredentials(String accessKey, String secretKey) {
+        if (accessKey == null) {
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/auth/BasicJCSCredentials.java]: Do Something like StringUtils.isNotBlank
AS "" "   " are all invalid strings

File: src/main/java/com/jcs/sbs/auth/BasicJCSCredentials.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            throw new IllegalArgumentException("Access key cannot be null.");
+        }
+        if (secretKey == null) {
+            throw new IllegalArgumentException("Secret key cannot be null.");
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/auth/BasicJCSCredentials.java]: Ditto!

File: src/main/java/com/jcs/sbs/auth/ConfigPropertiesCredentialsProvider.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    @Override
+    public JCSCredentials getCredentials() {
+
+        String accessKey = PropertiesReader.getProperty("ACCESS_KEY").trim();
+
+        String secretKey = PropertiesReader.getProperty("SECRET_KEY").trim();
+
+        if (Utils.isNullOrEmpty(accessKey) || Utils.isNullOrEmpty(secretKey)) {
+
+            throw new RuntimeException("Unable to load JCS credentials from config.properties file (" + accessKey
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/auth/ConfigPropertiesCredentialsProvider.java]: Use String.format

File: src/main/java/com/jcs/sbs/auth/EnvironmentVariableCredentialsProvider.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+     */
+    @Override
+    public JCSCredentials getCredentials() {
+        String accessKey = System.getenv("ACCESS_KEY");
+        String secretKey = System.getenv("SECRET_KEY");
+
+        accessKey = accessKey.trim();
+        secretKey = secretKey.trim();
+
+        if (Utils.isNullOrEmpty(accessKey) || Utils.isNullOrEmpty(secretKey)) {
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/auth/EnvironmentVariableCredentialsProvider.java]: Use StringUtils

File: src/main/java/com/jcs/sbs/auth/JCSCredentials.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package com.jcs.sbs.auth;
+
+/**
+ * Interface that provides access to Access and Secret keys
+ *
+ */
+public interface JCSCredentials {
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/auth/JCSCredentials.java:7]: Java docs on methods

File: src/main/java/com/jcs/sbs/auth/JCSCredentialsProviderChain.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        if (reuseLastProvider && lastUsedProvider != null) {
+            return lastUsedProvider.getCredentials();
+        }
+
+        for (JCSCredentialsProvider provider : credentialsProviders) {
+            try {
+                JCSCredentials credentials = provider.getCredentials();
+
+                if (credentials.getJCSAccessKey() != null && credentials.getJCSSecretKey() != null) {
+                    log.debug("Loading credentials from " + provider.toString());
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/auth/JCSCredentialsProviderChain.java]: Remove the debug line here

File: src/main/java/com/jcs/sbs/auth/JCSCredentialsProviderChain.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+                    lastUsedProvider = provider;
+                    return credentials;
+                }
+            } catch (Exception e) {
+                // Ignore any exceptions and move onto the next provider
+                log.debug("Unable to load credentials from " + provider.toString() + ": " + e.getMessage());
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/auth/JCSCredentialsProviderChain.java]: This is an exception. it should either be a warn or error

File: src/main/java/com/jcs/sbs/auth/JCSCredentialsProviderChain.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+            }
+        }
+
+        throw new RuntimeException("Unable to load JCS credentials from any provider in the chain");
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/auth/JCSCredentialsProviderChain.java]: Throw the exception as RuntimeException(e)

File: src/main/java/com/jcs/sbs/common/JCSHttpClient.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+                credentials.getJCSSecretKey(), headers);
+        authorization.addAuthorization(params);
+        String url = getUrl(params);
+        System.out.println(url);
+        CloseableHttpClient httpclient = HttpClients.createDefault();
+        HttpRequestBase httpGet = new HttpGet(url);
+
+        for (String header : headers.keySet()) {
+            httpGet.addHeader(header, headers.get(header));
+            ;
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/common/JCSHttpClient.java]: Remove this

File: src/main/java/com/jcs/sbs/common/JCSHttpClient.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        CloseableHttpResponse closeableHttpResponse = httpclient.execute(httpGet);
+
+        System.out.println("Response Code: " + closeableHttpResponse.getStatusLine().getStatusCode());
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/common/JCSHttpClient.java]: Remove this

File: src/main/java/com/jcs/sbs/common/JCSHttpClient.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+    /**
+     * returns the URL after adding query parameters.
+     * 
+     * @param params
+     *            Query parameters in the form of map that are to be present in
+     *            the API being called.
+     * @return Query parameters in the form of string in URL format.
+     */
+    private String getUrl(Map<String, String> params) {
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/common/JCSHttpClient.java:111]: Advisable to use string builders here

File: src/main/java/com/jcs/sbs/common/PropertiesReader.java [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+     * @param key
+     *            The key to be searched in config.properties file
+     * @return The value (if any) corresponding to given key
+     */
+    public static String getProperty(String key) {
+        try {
+            properties.load(input);
+            return properties.getProperty(key);
+        } catch (IOException e1) {
+            e1.printStackTrace();
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/common/PropertiesReader.java]: use loggers

File: src/main/java/com/jcs/sbs/common/Utils.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    /**
+     * This method finds protocol and host for given url.
+     * 
+     * @param url
+     *            The url from which protocol and host are to be found
+     * @return ProtocolAndHost object.
+     */
+    public static ProtocolAndHost getProtocolAndHost(String url) {
+        ProtocolAndHost protocolAndHost = new ProtocolAndHost();
+        Pattern r = Pattern.compile("(http[s]?)://((?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*(),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+)");
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/common/Utils.java]: Use a string constant for the pattern

File: src/main/java/com/jcs/sbs/common/Utils.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        Matcher url_parts = r.matcher(url);
+        if (!url_parts.find()) {
+            protocolAndHost.setProtocol("");
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/common/Utils.java]: What happens when protocol and host are empty ?

File: src/main/java/com/jcs/sbs/common/Utils.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    /**
+     * This method checks if the given string value is null or empty.
+     * 
+     * @param value
+     *            The string value that is to be tested
+     * @return Boolean value. True if string is null or empty, false otherwise.
+     */
+    public static boolean isNullOrEmpty(String value) {
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/common/Utils.java]: Remove this. Use StringUtils else where

File: src/main/java/com/jcs/sbs/main/JCStry.java [hunk 18]
```diff
@@ -0,0 +1,110 @@
+package com.jcs.sbs.main;
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/main/JCStry.java:1]: Remove this test file from SDK

File: src/main/java/com/jcs/sbs/model/DescribeSnapshotsResult.java [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+    private List<Snapshot> snapshots;
+
+    private String nextToken;
+
+    /**
+     * @return
+     */
+    public List<Snapshot> getSnapshots() {
+        if (snapshots == null) {
+            snapshots = new ArrayList<Snapshot>();
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/model/DescribeSnapshotsResult.java]: Why not initialize snapshots = new ArrayList<Snapshot>(); at the time of object creation ??
Follow this pattern elsewhere

File: src/main/java/com/jcs/sbs/model/DescribeVolumesRequest.java [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+    public DescribeVolumesRequest(List<String> volumeIds) {
+        setVolumeIds(volumeIds);
+    }
+
+    /**
+     * @return
+     */
+    public List<String> getVolumeIds() {
+        if (volumeIds == null) {
+            volumeIds = new ArrayList<String>();
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/model/DescribeVolumesRequest.java]: Ditto explained above

File: src/main/java/com/jcs/sbs/service/impl/JCSComputeClient.java [hunk 21]
```diff
@@ -0,0 +1,510 @@
...
+                EntityUtils.consume(entity);
+                CreateVolumeResult createVolumeResult = new CreateVolumeResult().withVolume(volume);
+                createVolumeResult.setRequestId(createVolumeResponse.getRequestId());
+                createVolumeResult.setXml(content);
+                return createVolumeResult;
+
+            } else {
+                HttpEntity entity = response.getEntity();
+                String content = EntityUtils.toString(entity);
+                System.out.println(content);
```
Comment (sssouvik) [inline src/main/java/com/jcs/sbs/service/impl/JCSComputeClient.java]: Remove this

File: src/test/java/com/jcs/sbs/service/JCSComputeTest.java [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+import com.jcs.sbs.model.DeleteSnapshotResult;
+import com.jcs.sbs.model.DeleteVolumeRequest;
+import com.jcs.sbs.model.DeleteVolumeResult;
+import com.jcs.sbs.model.DescribeSnapshotsRequest;
+import com.jcs.sbs.model.DescribeSnapshotsResult;
+import com.jcs.sbs.model.DescribeVolumesRequest;
+import com.jcs.sbs.model.DescribeVolumesResult;
+import com.jcs.sbs.service.impl.JCSComputeClient;
+
+public class JCSComputeTest {
```
Comment (sssouvik) [inline src/test/java/com/jcs/sbs/service/JCSComputeTest.java:38]: Remove this from SDK

File: src/main/java/com/jcs/sbs/auth/JCSCredentialsProviderChain.java (hunk 8 above)
Comment (utsav-deep-rjil) [reply src/main/java/com/jcs/sbs/auth/JCSCredentialsProviderChain.java]: I think user should be informed which credentials is actually being used in SDK. If the user had set the credentials at all the places (config.properties, environment variables and system properties) but all are different, then this info will be helpful.
BTW, I am changing the 'debug' into 'info'

File: src/main/java/com/jcs/sbs/model/Snapshot.java [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+	 * @return
+	 */
+	public Snapshot withEncryption(Boolean encrypted) {
+        setEncrypted(encrypted);
+        return this;
+    }
+    
+
+	@Override
+    public String toString() {
```
Comment (aggarwalchirag) [inline src/main/java/com/jcs/sbs/model/Snapshot.java:250]: starttime and status are not captured in toString

File: src/main/java/com/jcs/sbs/model/Snapshot.java [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+            sb.append("VolumeId: " + getVolumeId() + ",");
+        if (getSize() != null)
+            sb.append("Size: " + getSize() + ",");
+        if (getEncrypted() != null)
+            sb.append("Encrypted: " + getEncrypted() + ",");
+        sb.append("}");
+        return sb.toString();
+    }
+
+    @Override
```
Comment (aggarwalchirag) [inline src/main/java/com/jcs/sbs/model/Snapshot.java:267]: here also, starttime and status are not used

File: src/main/java/com/jcs/sbs/model/Snapshot.java [hunk 25]
```diff
@@ -0,0 +1,259 @@
...
+                && other.getSize().equals(this.getSize()) == false)
+            return false;
+        if (other.getEncrypted() == null ^ this.getEncrypted() == null)
+            return false;
+        if (other.getEncrypted() ^ this.getEncrypted())
+            return false;
+               
+        return true;
+    }
+
```
Comment (aggarwalchirag) [inline src/main/java/com/jcs/sbs/model/Snapshot.java:308]: here as well

File: src/main/java/com/jcs/sbs/model/Volume.java [hunk 26]
```diff
@@ -0,0 +1,275 @@
...
+	/**
+	 * @param volumeType
+	 */
+	public void setVolumeType(VolumeType volumeType) {
+		this.volumeType = volumeType;
+	}
+	/**
+	 * @param volumeType
+	 * @return
+	 */
```
Comment (aggarwalchirag) [inline src/main/java/com/jcs/sbs/model/Volume.java]: where is attachment set information ?

File: src/main/java/com/jcs/sbs/model/Volume.java [hunk 27]
```diff
@@ -0,0 +1,275 @@
...
+	public Volume withVolumeType(VolumeType volumeType) {
+        setVolumeType(volumeType);
+        return this;
+    }
+
```
Comment (aggarwalchirag) [inline src/main/java/com/jcs/sbs/model/Volume.java:290]: all members are not used here

File: src/main/java/com/jcs/sbs/model/Volume.java [hunk 28]
```diff
@@ -0,0 +1,275 @@
...
+	
+
+	private String volumeId;
+
+    private Integer size;
+
+    private VolumeType volumeType = VolumeType.standard;
+    
+    private String snapshotId;
+
```
Comment (aggarwalchirag) [inline src/main/java/com/jcs/sbs/model/Volume.java]: why is this initialized to true ?

**Review by aggarwalchirag — CHANGES_REQUESTED** [review]

File: src/main/java/com/jcs/sbs/model/DescribeSnapshotsRequest.java [hunk 29]
```diff
@@ -30,57 +39,51 @@ public DescribeSnapshotsRequest(List<String> snapshotsIds) {
...
+     * Sets the list of snapshot IDs that are to be described. To describe all
+     * the snapshots, there is no need to set snapshotIds
      * 
      * @param snapshotIds
      *            The IDs of the snapshot that are to be described.
      */
     public void setSnapshotIds(Collection<String> snapshotIds) {
         if (snapshotIds == null) {
-            this.snapshotIds = null;
+            this.snapshotIds = new ArrayList<String>();
```
Comment (aggarwalchirag) [inline src/main/java/com/jcs/sbs/model/DescribeSnapshotsRequest.java:69]: why do you need this. You initialized snapshots in the constructor itself. I can imagine a case, where someone calls up this function twice. In that scenario the change makes sense. But then you should handle it in withSnapshotIds as well.

**Review by aggarwalchirag — CHANGES_REQUESTED** [review]

**Review by aggarwalchirag — APPROVED** [review]
The code looks much cleaner now. @utsav-deep-rjil : Good Job.

**Pull request merged**